rfcs/20191016-dlpack-support.md

```diff
@@ -0,0 +1,93 @@
+
+A working version of dlpack integration has been released as a package by wmjlyjemaine@gmail.com and @VoVAllen here:
+https://github.com/VoVAllen/tf-dlpack/issues/3
+
+This proposal would leverage that solution and integrate it into TF so that the operations could be performed natively.
+
+User experience
+We plan to release a python package tfdlpack, containing two APIs:
+
+to_dlpack: Given a tensorflow tensor, return a DLPack tensor contain.
+from_dlpack: Given a DLPack-compatible python capsule, return a tensorflow tensor.
+Example codes of converting a Tensorflow tensor to Torch tensor using DLPack:
+```python
+import numpy as np
+import tensorflow as tf
+import torch.utils.dlpack as thdlpack
+import tfdlpack
+
+t1 = tf.constant([1, 2, 3], dtype=np.float32)
+dlpack = tfdlpack.to_dlpack(t1)  # tf tensor -> dlpack
+t2 = thdlpack.from_dlpack(dlpack)  # dlpack -> th tensor
+print(t2)
+dlpack = thdlpack.to_dlpack(t2)  # th tensor -> dlpack
+t3 = tfdlpack.from_dlpack(dlpack)  # dlpack -> tf tensor
+print(t3)
+```
+You will find that t1, t2 and t3 all have the same values, shape, and device contexts.
+Package dependency: tensorflow>=2.0
+
+How it works?
+The first design consideration is that we want to avoid any modification to the main Tensorflow library, so to get around the potential long delay of PR, code review, and release cycle of Tensorflow main package. Inspired by the solution from https://github.com/tobegit3hub/tftvm, we decide to implement the functionality as two custom tensor ops: to_dlpack and from_dlpack.
+
+Besides, we want this feature to be plugged into other projects quite easily. For example, any project that relies on this feature is able to run without compiling against Tensorflow's header files. Not only that an extra dependency usually means extra effort, but also that such maintenance is repetitive and should be handled by the feature developer (i.e., us) alone. To this end, we have an idea of releasing it as a python package. However, the question is how to invoke the two custom tensor ops in python? The challenge is that Tensorflow's custom op interface has a limited support of argument and return types, while to_dlpack and from_dlpack should have an argument/return type of DLPack object. We work around this by encoding the address of an DLPack object as an integer, so it can be accepted/returned by the custom op interface. Then, we decode it in python or C depending on whether we return it (to_dlpack) or consume it (from_dlpack).
```

Comment:
Yes. We hope this could be part of tensorflow. Currently we are using lots of workaround to prototype, which is not ideal. However, we need input from tensorflow people to see whether there's a more elegant way to do this.

Reply:
I've revised the RFC to be more clear on this.